Fetch data asynchronously

glassez opened the pull request:
This (together with #18010) should prevent UI from being freezed when updates some data fetching from libtorrent when it does some hard work.
I tried to touch on all cases when it comes to regular auto-updates. There are still untouched cases when data is fetched on demand, for example, exporting a .torrent file or magnet link, etc. The Web API requires additional refactoring so that a similar approach can be applied there. This will follow further.

thalieht commented:
I can't find any problems. I don't have any torrents with web seeds to test that part.

xavier2k6 commented:
>I can't find any problems.

same.

>I don't have any torrents with web seeds to test that part.

Do you mean torrents that have `https` & appear under `http sources` or `ws/wss`?

thalieht commented:
>>I don't have any torrents with web seeds to test that part.
>
>Do you mean torrents that appear under http sources

Yes.

xavier2k6 commented:
> > > I don't have any torrents with web seeds to test that part.
> > 
> > 
> > Do you mean torrents that appear under http sources
> 
> Yes.

👍 

You can test with gimp

https://www.gimp.org/downloads/

<img width="629" alt="gimp 2 10" src="https://user-images.githubusercontent.com/42386382/202874037-ee0cab3c-d4e5-4a97-aabf-7ccd71f2fde1.png">

thalieht commented:
Thanks @xavier2k6.
Uhh web seeds are messed up in master and here. 
Deleting one web seed deletes them all.
On qBt restart all web seeds are gone
An additional problem with this PR is that i can't add a new web seed.

xavier2k6 commented:
>Deleting one web seed deletes them all.

>i can't add a new web seed.

I'm not seeing that exactly but what I am seeing with this PR is that you will have to either add/delete the web seed a 2nd time for the action to be reflected in http sources tab or close qBittorrent & re-open after adding/deleting a web seed.?!

thalieht commented:
>I'm not seeing that exactly but what I am seeing with this PR is that you will have to either add/delete the web seed a 2nd time for the action to be reflected in http sources tab & either select a different torrent/go back or close qBittorrent & re-open after adding/deleting a web seed.?!

Hmm, yes sometimes i get these and sometimes what i described above.

thalieht commented:
Could this web-seeds problem be libtorrent related? I went all the way back to qbt 4.4.0rc1 commit and problem is still present with libt 2.0.8 but there is no such bug in official 4.4.5 with libt 2.0.7.

glassez commented:
> Could this web-seeds problem be libtorrent/Qt related? I went all the way back to qbt 4.4.0rc1 commit and problem is still present with libt 2.0.8/Qt 6.4.0 but there is no such bug in official 4.4.5 with libt 2.0.7/Qt 6.3.0.

Don't think so.
I've found and fixed my mistake in concurrency handling when adding/removing web seeds.
>On qBt restart all web seeds are gone

This was because #18033 still hadn't been merged into the master. Now I have merged it and rebased this PR on top of it.

thalieht commented:
https://libtorrent.org/reference-Torrent_Handle.html#remove-url-seed-url-seeds-add-url-seed
>Note that URLs that fails may be removed automatically from the list.

I think this is what happens... they slowly start disappearing from the list a few at a time in the Gimp torrent and only when torrent is not paused.

In that case, no more problems with this PR.

glassez commented:
> > Note that URLs that fails may be removed automatically from the list.
> 
> I think this is what happens... they slowly start disappearing from the list a few at a time in the Gimp torrent and only when torrent is not paused.

Yes, the processing of web seeds is rather curious. As far as I understand, they can be automatically changed (in case of HTTP redirection) and even new ones are added in some cases. So I think it is possible to objectively evaluate the correctness of adding/removing web seeds only on a stopped torrent. Otherwise you may be confusing by such "behind the scenes" actions.

stalkerok commented:
To test web seeds, you can still use LibreOffice, they also have many links, but by default, based on geolocation, only one web seed is added to the torrent file.
https://download.documentfoundation.org/libreoffice/stable/7.4.2/win/x86_64/LibreOffice_7.4.2_Win_x64.msi.mirrorlist

sledgehammer999 commented on src/gui/properties/peerlistwidget.cpp:
```diff
@@ -433,7 +443,7 @@ void PeerListWidget::updatePeer(const BitTorrent::Torrent *torrent, const BitTor
-        [this] (const int row, const int column, const QString &displayData
+        [this](const int row, const int column, const QString &displayData
                 , const QVariant &underlyingData, const Qt::Alignment textAlignmentData = {}
                 , const QString &toolTip = {})
```
I think the 2 lines above need to have a space deleted too. (to align correctly)

glassez replied:
>to align correctly

What is correct alignment? Currently they are shifted by several count of indentations (so their overall indentation is power of 4 spaces).

sledgehammer999 replied:
I think it is customary to begin one character after the opening parentheses in this case.

glassez replied:
> I think it is customary to begin one character after the opening parentheses in this case.

We don't have strict rules about this. Therefore, I would prefer not to dwell on this here, but to discuss it separately, if that's what you care about.
I'll just say right away that such "conditional" alignments make the code terrible in some (very frequent) cases, since it can leave too much unjustifiably unused space on the left, unlike the fixed-indent method.

glassez commented on src/base/bittorrent/torrent.h:
```diff
@@ -128,7 +129,7 @@ namespace BitTorrent
         static const int MAX_SEEDING_TIME;
 
-        virtual ~Torrent() = default;
+        using QObject::QObject;
```
It allows you to use base class constructors (e.g. `Torrent(QObject *parent)` is accessible now).

glassez commented on src/base/bittorrent/torrentimpl.cpp:
```diff
@@ -609,50 +610,86 @@ QVector<QUrl> TorrentImpl::urlSeeds() const
+
+            session->invoke([session, thisTorrent, addedUrlSeeds]
+            {
+                if (!thisTorrent)
```
I'd rather not mix them. These are conditions of various kinds. The first relates to the concurrency handling, and the second to the basic logic of the handler.

sledgehammer999 replied:
Then maybe this?
```c++
if (addedUrlSeeds.isEmpty())
    return;
```

as a way to avoid one indentation level. I don't insist on this. You decide.

glassez replied:
> as a way to avoid one indentation level. I don't insist on this. You decide.

I don't see any problems with readability in this case, so I would prefer to leave it as it is.

glassez replied:
In addition, having it in the way you proposed will give another reason for future "optimizers" to merge these two diverse conditions together: `if (!thisTorrent || addedUrlSeeds.isEmpty()) return;`.

glassez commented on src/base/bittorrent/torrentimpl.cpp:
```diff
@@ -2408,6 +2445,163 @@ nonstd::expected<void, QString> TorrentImpl::exportToFile(const Path &path) cons
+    {
+        QVector<qreal> result;
+        if (!torrentInfo.isValid())
+            return result;
```
> Just a suggestion. Move down the declaration of `result` just before the `try` keyword and here do `return {};` instead.

Ok.

glassez commented on src/base/bittorrent/torrentimpl.cpp:
```diff
@@ -2408,6 +2445,163 @@ nonstd::expected<void, QString> TorrentImpl::exportToFile(const Path &path) cons
+            std::vector<int64_t> fp;
+            nativeHandle.file_progress(fp, lt::torrent_handle::piece_granularity);
+
+            const int count = torrentInfo.filesCount();
```
> This value is already available in the `filesCount` variable.

👍

glassez commented on src/base/bittorrent/torrentimpl.cpp:
```diff
@@ -2408,6 +2445,163 @@ nonstd::expected<void, QString> TorrentImpl::exportToFile(const Path &path) cons
+        try
+        {
+            std::vector<int64_t> fp;
+            nativeHandle.file_progress(fp, lt::torrent_handle::piece_granularity);
```
Ok, changed.
